source/extended-json.rst

```diff
@@ -20,233 +20,413 @@ Extended JSON
+to parsing JSON numbers:
+
+* If the number is a *non-integer*, parsers SHOULD interpret it as BSON Double.
+
+* If the number is an *integer*, parsers SHOULD interpret it as being of the
+  smallest BSON integer type that can represent the number exactly.  If a parser
+  is unable to represent the number exactly as an integer (e.g.  a large 64-bit
+  number on a 32-bit platform), it MUST interpret it as a BSON Double even if
+  this results in a loss of precision.  The parser MUST NOT interpret it as a
+  BSON String containing a decimal representation of the number.
+
+Generators
+----------
+
+An Extended JSON generator (hereafter just "generator") produces strings in an
+Extended JSON format.
+
+A generator MUST allow users to produce strings in either the Canonical
+Extended JSON format or the Relaxed Extended JSON format.  If generators
+provide a default format, the default SHOULD be the Relaxed Extended JSON
+format.
+
+A generator MAY be capable of exporting strings that adhere to other
+formats, such as Legacy Extended JSON formats.
+
+A generator SHOULD support at least 100 `levels of nesting`_ in a BSON
+document.
+
+Transforming BSON
+.................
+
+Given a BSON document (e.g. a buffer of bytes meeting the requirements of the
+BSON specification), a generator MUST use the corresponding JSON values or
+Extended JSON type wrapper objects for the BSON type given in the Extended JSON
+`Conversion table`_ for the desired format.  When transforming a BSON document
+into Extended JSON text, a generator SHOULD emit the JSON keys and values in
+the same order as given in the BSON document.
+
+Transforming Language-Native data
+.................................
+
+Given language-native data (e.g. type primitives, container types, classes,
+etc.), if there is a semantically-equivalent BSON type for a given
+language-native type, a generator MUST use the corresponding JSON values or
+Extended JSON type wrapper objects for the BSON type given in the Extended JSON
+`Conversion table`_ for the desired format.  For example, a Python ``datetime``
+object must be represented the same as a BSON datetime type.  A generator
+SHOULD error if a language-native type has no semantically-equivalent BSON
+type.
+
+Format and Method Names
+.......................
+
+The following format names SHOULD be used for selecting formats for generator
+output:
+
+* ``canonicalExtendedJSON`` (references Canonical Extended JSON as described in
+  this specification)
+
+* ``relaxedExtendedJSON`` (references Relaxed Extended JSON as described in
+  this specification)
+
+* ``legacyExtendedJSON`` (if supported: references Legacy Extended JSON,
+  with implementation-defined behavior)
+
+Generators MAY use these format names as part of function/method names or MAY
+use them as arguments or constants, as needed.
+
+If a generator provides a generic `to_json` or `to_extended_json` method, it
+MUST default to producing Relaxed Extended JSON.
```

Comment:
This is the approach we're planning for libbson in [CDRIVER-2208](https://jira.mongodb.org/browse/CDRIVER-2208).